WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/BlockEditorPage.java

```diff
@@ -27,12 +36,113 @@ public BlockEditorPage() {
+        return this;
+    }
+
+    public BlockEditorPage switchToHtmlMode() {
+        openActionBarOverflowOrOptionsMenu(ApplicationProvider.getApplicationContext());
+        clickOn(onView(withText(R.string.menu_html_mode)));
+        return this;
+    }
+
+    public BlockEditorPage switchToVisualMode() {
+        openActionBarOverflowOrOptionsMenu(ApplicationProvider.getApplicationContext());
+        clickOn(onView(withText(R.string.menu_visual_mode)));
+        return this;
+    }
+
+    public BlockEditorPage addPostSettings(String categoryName, String tagName) {
+        openPostSetting();
+        addCategory(categoryName);
+        addTag(tagName);
+        pressBack();
+        return this;
+    }
+
+    public void openPostSetting() {
+        openActionBarOverflowOrOptionsMenu(ApplicationProvider.getApplicationContext());
+        clickOn(onView(withText(R.string.post_settings)));
+    }
+
+    public void addCategory(String category) {
+        clickOn(onView(withId(R.id.post_categories_container)));
+        clickOn(onView(withText(category)));
+        pressBack();
+    }
+
+    public void addTag(String tag) {
+        clickOn(onView(withId(R.id.post_tags_container)));
+        ViewInteraction tagsField = onView(withId(R.id.tags_edit_text));
+        populateTextField(tagsField, tag);
+        pressBack();
+    }
+
+    public BlockEditorPage publish() {
+        clickPublish();
+        confirmPublish();
+        return this;
+    }
+
+    public BlockEditorPage clickPublish() {
+        clickOn(onView(withResourceName("menu_primary_action")));
+        return this;
+    }
+
+    public BlockEditorPage confirmPublish() {
+        clickOn(onView(withText(R.string.publish_now)));
+        dismissBloggingRemindersAlertIfNeeded();
+        return this;
+    }
+
+    public void dismissBloggingRemindersAlertIfNeeded() {
+        ViewInteraction bloggingRemindersAlertTitle = onView(withText(R.string.set_your_blogging_reminders_title));
+
+        if (waitForElementToBeDisplayedWithoutFailure(bloggingRemindersAlertTitle)) {
+            bloggingRemindersAlertTitle.perform(swipeDown());
+        }
+    }
+
+    public void verifyPostPublished() {
+        assertTrue("'Post published' toast was not displayed",
+                waitForElementToBeDisplayedWithoutFailure(onView(withText(R.string.post_published))));
+    }
+
+    public BlockEditorPage verifyPostSettings(String categoryName, String tagName) {
+        assertTrue("Category is not present in Post confirmation panel",
+                waitForElementToBeDisplayedWithoutFailure(onView(withText(categoryName))));
+        assertTrue("Tag is not present in Post confirmation panel",
+                waitForElementToBeDisplayedWithoutFailure(onView(withText(tagName))));
+        return this;
+    }
+
+    public BlockEditorPage verifyPostElementText(String postText) {
+        assertTrue("Expected text is not displayed",
```

Comment:
Thinking of possible failures, this looks for the text to appear in visual mode. If tapping on visual mode does not update the page and still shows HTML mode (or if the page is slow to update), would this still pass?

Reply:
Excellent point, @jostnes! 
I reproduced your scenario by removing the `.switchToVisualMode()` from the test and it fails because the assertion will look for an Element with exact text and in`.verifyPostElementText(mPostText)` we are checking for a substring of the original HTML post `enterParagraphText(mHtmlPost).`
```
String mPostText = "Lorem ipsum dolor sit amet, consectetur adipiscing elit.";
String mHtmlPost = "<!-- wp:paragraph -->\n"
                       + "<p>" + mPostText + "</p>\n"
                       + "<!-- /wp:paragraph -->"
                       ...
```

Reply:
Thanks for checking! All good then.